### PR TITLE
Refactor sync controller into separate traits

### PR DESCRIPTION
**Pull request by LukasGasior1**

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+    }
+  }
+
+  private def startingFastSync: Receive = handlePeerUpdates orElse {
+    case RetryStart => startFastSync()
+  }
+
+  private def startFastSync(syncState: SyncState): Unit = {
+    scheduler.schedule(0.seconds, printStatusInterval, self, PrintStatus)
+    context become new SyncingHandler(syncState).receive
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: should context be combined with
`handlePeerUpdates orElse` ?

**Comment by adamsmo** [comment]
other than question LGTM

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+      peersUsedToChooseTarget.foreach { case (peer, Handshaked(status, _, _)) =>
+        peer ! PeerActor.Subscribe(Set(BlockHeaders.code))
+        peer ! PeerActor.SendMessage(GetBlockHeaders(Right(status.bestHash), 1, 0, reverse = false))
+      }
+      log.info("Asking {} peers for block headers", peersUsedToChooseTarget.size)
+      val timeout = scheduler.scheduleOnce(peerResponseTimeout, self, BlockHeadersTimeout)
+      context become waitingForBlockHeaders(peersUsedToChooseTarget.keySet, Map.empty, timeout)
+    } else {
+      log.warning("Cannot start fast sync, not enough peers to download from. Scheduling retry in {}", startRetryInterval)
+      scheduleStartRetry(startRetryInterval)
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: Should we do here `context become startingFastSync`? I'm not sure whether we can handle the `RetryStart` message if that's not the case, I wasn't able to start our node without this line.

File: src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala [hunk 3]
```diff
@@ -23,11 +21,13 @@ trait RegularSyncController {
 
   import Config.FastSync._
 
-  def regularSync(): Receive = {
-
-    case StartSyncing =>
-      askForHeaders()
+  def startRegularSync(): Unit = {
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala]: Should we do here `appStateStorage.fastSyncDone()` as we previously did when handling the `StartRegularSync` message? If that's not the case we are able to run first the regular sync and then the fast sync, which I think we agreed that we shouldn't allow in #90.

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala (hunk 1 above)
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala]: It's already in `SyncingHandlerer.receive`

**Comment by LukasGasior1** [comment]
Thanks @ntallar for pointing that out. Fixed now.

**Comment by adamsmo** [comment]
I tested this branch and it looks like fast sync does not work when running sbt run with empty db

**Pull request merged**